packages/omgidl-parser/src/IdlNodes/ReferenceTypeIdlNode.ts

```diff
@@ -0,0 +1,204 @@
+    }
+    return this.astNode.type;
+  }
+
+  get isComplex(): boolean {
+    return this.needsResolution ? this.parent().isComplex : false;
+  }
+
+  get isArray(): boolean | undefined {
+    let isArray = this.astNode.isArray;
+
+    if (this.needsResolution) {
+      const parent = this.parent();
+      if (parent instanceof StructIdlNode) {
+        return this.astNode.isArray;
+      }
+      if (parent.isArray != undefined) {
+        isArray = parent.isArray;
+      }
+    }
+    return isArray;
+  }
+
+  get arrayLengths(): number[] | undefined {
+    const arrayLengths = this.astNode.arrayLengths ? [...this.astNode.arrayLengths] : [];
+    if (this.needsResolution) {
+      const parent = this.parent();
+      if (!(parent instanceof StructIdlNode) && parent.arrayLengths) {
+        arrayLengths.push(...parent.arrayLengths);
+      }
+    }
+    const finalArrayLengths: number[] = [];
+    for (const arrayLength of arrayLengths) {
+      if (typeof arrayLength === "object") {
+        finalArrayLengths.push(this.getConstantNode(arrayLength.name).value as number);
+        continue;
+      }
+      finalArrayLengths.push(arrayLength);
+    }
+    return finalArrayLengths.length > 0 ? finalArrayLengths : undefined;
+  }
+
+  get arrayUpperBound(): number | undefined {
+    let arrayUpperBound = undefined;
+    if (this.needsResolution) {
+      const parent = this.parent();
+      if (parent instanceof StructIdlNode) {
+        return undefined;
+      }
+      arrayUpperBound = parent.arrayUpperBound;
+    }
+    // prioritize local arrayUpperBound
+    if (this.astNode.arrayUpperBound != undefined) {
+      arrayUpperBound = this.astNode.arrayUpperBound;
+    }
+    if (typeof arrayUpperBound === "object") {
+      arrayUpperBound = this.getConstantNode(arrayUpperBound.name).value as number;
+    }
+    return arrayUpperBound;
+  }
+
+  get upperBound(): number | undefined {
+    let upperBound = undefined;
+    if (this.needsResolution) {
+      const parent = this.parent();
+      if (parent instanceof StructIdlNode) {
+        return undefined;
+      }
+      return parent.upperBound;
+    }
+    if (this.astNode.upperBound != undefined) {
+      upperBound = this.astNode.upperBound;
+    }
+    if (typeof upperBound === "object") {
+      upperBound = this.getConstantNode(upperBound.name).value as number;
+    }
+    return upperBound;
+  }
+
+  get annotations(): BaseASTNode["annotations"] {
+    let annotations = undefined;
+    if (this.needsResolution) {
+      const parent = this.parent();
+      if (parent instanceof StructIdlNode) {
+        return undefined;
+      }
+      if (parent.annotations != undefined) {
+        annotations = { ...parent.annotations };
+      }
+    }
+    if (this.astNode.annotations != undefined) {
+      annotations = { ...annotations, ...this.astNode.annotations };
+    }
+    return annotations;
+  }
+
+  getValidFieldReference(typeName: string): PossibleParentNode {
```

Comment:
Comment? From looking at the name and signature I cannot figure out what this function is for. It seems like it is looking for a reference to a type name (is it resolving that reference?) - why does it return something that is a "possible parent"?

Reply:
renamed this and `parent`-related variables to call them `TypeRef`since they're more just references to another type than an actual parent child relationship. Hopefully that helps clear things up in addition to the comment